core/src/main/java/com/linecorp/armeria/server/RedirectService.java

```diff
@@ -0,0 +1,185 @@
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.server;
+
+import static java.util.Objects.requireNonNull;
+
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Function;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import com.linecorp.armeria.common.HttpHeaderNames;
+import com.linecorp.armeria.common.HttpHeaders;
+import com.linecorp.armeria.common.HttpRequest;
+import com.linecorp.armeria.common.HttpResponse;
+import com.linecorp.armeria.common.HttpStatus;
+
+/**
+ * A {@link Service} that implements an HTTP redirection.
+ *
+ * <p>This class will redirect requests from its bound {@link Service} path pattern
+ * to a new location path pattern. Currently only these path patterns are supported:
+ * <ul>
+ *   <li>{@code /new} (no path parameters)</li>
+ *   <li>{@code /new/{var1}} (curly-brace style)</li>
+ *   <li>{@code /new/:var1/new/:var2} (colon style)</li>
+ *   <li>{@code http://host/new} (full url no path parameters)</li>
+ *   <li>{@code http://host/new/{var1}} (full url and curly-brace style param)</li>
+ *   <li>{@code http://host/new/:var1/new/:var2} (full url and colon style param)</li>
+ * </ul>
+ * The {@link RedirectService} will return {@link HttpStatus#TEMPORARY_REDIRECT 307 Temporary Redirect}
+ * by default.
+ * <pre>{@code
+ * ServerBuilder sb = ...;
+ * sb.service("/old/", new RedirectService("/new"));
+ * sb.service("/old/{var}", new RedirectService("/new/{var}");
+ * sb.service("/old/{var}", new RedirectService(ctx -> "/new/" + ctx.pathParam("var"));
+ *
+ * sb.service("/old/{var}", new RedirectService(HttpStatus.MOVED_PERMANENTLY, "/new/{var}");
+ * sb.service("/old/{var}", new RedirectService("http://user:name@localhost:8080/search/{var}");
+ * }</pre>
+ */
+public class RedirectService extends AbstractHttpService {
+
+    private static final Pattern VALID_DEFAULT_URI_PATTERN = Pattern
+            .compile("(?:(?:^https?:/{2}(?:([^:]+:)?[^:@]+@)?[^:]+)(?::[0-9]{1,5})?)?" +
+                    "(?:/[^/{}:]+|/:[^/{}]+|/\\{[^/{}]+})+/?");
+    private static final Pattern PATTERN_PARAMS_START = Pattern.compile("/:|/\\{");
+
+    private final HttpStatus httpStatus;
+    private final Function<? super ServiceRequestContext, String> locationFunction;
+    private Set<String> paramNames;
+
+    /**
+     * Creates a new Redirect {@link Service} that redirects to the location constructed with the specified
+     * {@code locationPattern}.
+     *
+     * @param locationPattern a {@link PathMapping} string that the bound {@link Service} will redirect to.
```

Comment:
`a {@link PathMapping} string that ...` -> `the location pattern that is used to generate a redirect location`